# TV Script Generation

In this project, you'll generate your own [Seinfeld](https://en.wikipedia.org/wiki/Seinfeld) TV scripts using RNNs.  You'll be using part of the [Seinfeld dataset](https://www.kaggle.com/thec03u5/seinfeld-chronicles#scripts.csv) of scripts from 9 seasons.  The Neural Network you'll build will generate a new ,"fake" TV script, based on patterns it recognizes in this training data.

## Get the Data

The data is already provided for you in `./data/Seinfeld_Scripts.txt` and you're encouraged to open that file and look at the text. 
>* As a first step, we'll load in this data and look at some samples. 
* Then, you'll be tasked with defining and training an RNN to generate a new script!

In [43]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# load in data
import helper
data_dir = './data/Seinfeld_Scripts.txt'
text = helper.load_data(data_dir)

## Explore the Data
Play around with `view_line_range` to view different parts of the data. This will give you a sense of the data you'll be working with. You can see, for example, that it is all lowercase text, and each new line of dialogue is separated by a newline character `\n`.

In [44]:
view_line_range = (0, 10)

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
import numpy as np

print('Dataset Stats')
print('Roughly the number of unique words: {}'.format(len({word: None for word in text.split()})))

lines = text.split('\n')
print('Number of lines: {}'.format(len(lines)))
word_count_line = [len(line.split()) for line in lines]
print('Average number of words in each line: {}'.format(np.average(word_count_line)))

print()
print('The lines {} to {}:'.format(*view_line_range))
print('\n'.join(text.split('\n')[view_line_range[0]:view_line_range[1]]))

Dataset Stats
Roughly the number of unique words: 46367
Number of lines: 109233
Average number of words in each line: 5.544240293684143

The lines 0 to 10:
jerry: do you know what this is all about? do you know, why were here? to be out, this is out...and out is one of the single most enjoyable experiences of life. people...did you ever hear people talking about we should go out? this is what theyre talking about...this whole thing, were all out now, no one is home. not one person here is home, were all out! there are people trying to find us, they dont know where we are. (on an imaginary phone) did you ring?, i cant find him. where did he go? he didnt tell me where he was going. he must have gone out. you wanna go out you get ready, you pick out the clothes, right? you take the shower, you get all ready, get the cash, get your friends, the car, the spot, the reservation...then youre standing around, what do you do? you go we gotta be getting back. once youre out, you wanna get back! y

---
## Implement Pre-processing Functions
The first thing to do to any dataset is pre-processing.  Implement the following pre-processing functions below:
- Lookup Table
- Tokenize Punctuation

### Lookup Table
To create a word embedding, you first need to transform the words to ids.  In this function, create two dictionaries:
- Dictionary to go from the words to an id, we'll call `vocab_to_int`
- Dictionary to go from the id to word, we'll call `int_to_vocab`

Return these dictionaries in the following **tuple** `(vocab_to_int, int_to_vocab)`

In [45]:
import problem_unittests as tests
from collections import Counter
def create_lookup_tables(text):
    """
    Create lookup tables for vocabulary
    :param text: The text of tv scripts split into words
    :return: A tuple of dicts (vocab_to_int, int_to_vocab)
    """
    # TODO: Implement Function
    vocab = [word for word in text]
    vocab_count = Counter(vocab)
    vocab_to_int = {word[0]:idx for idx, word in enumerate(list(vocab_count.most_common()))}
    int_to_vocab = {idx:word for word, idx in vocab_to_int.items()}
    # return tuple
    return (vocab_to_int, int_to_vocab)


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_create_lookup_tables(create_lookup_tables)

Tests Passed


### Tokenize Punctuation
We'll be splitting the script into a word array using spaces as delimiters.  However, punctuations like periods and exclamation marks can create multiple ids for the same word. For example, "bye" and "bye!" would generate two different word ids.

Implement the function `token_lookup` to return a dict that will be used to tokenize symbols like "!" into "||Exclamation_Mark||".  Create a dictionary for the following symbols where the symbol is the key and value is the token:
- Period ( **.** )
- Comma ( **,** )
- Quotation Mark ( **"** )
- Semicolon ( **;** )
- Exclamation mark ( **!** )
- Question mark ( **?** )
- Left Parentheses ( **(** )
- Right Parentheses ( **)** )
- Dash ( **-** )
- Return ( **\n** )

This dictionary will be used to tokenize the symbols and add the delimiter (space) around it.  This separates each symbols as its own word, making it easier for the neural network to predict the next word. Make sure you don't use a value that could be confused as a word; for example, instead of using the value "dash", try using something like "||dash||".

In [46]:
def token_lookup():
    """
    Generate a dict to turn punctuation into a token.
    :return: Tokenized dictionary where the key is the punctuation and the value is the token
    """
    # TODO: Implement Function
    token_dict = {'.': '||Period||',
                  ',': '||Comma||',
                  '"': '||Quotation_Mark||',
                  ';': '||Semicolon||',
                  '!': '||Exclamation_Mark||',
                  '?': '||Question_Mark||',
                  '(': '||Left_Parentheses||',
                  ')': '||Right_Parentheses||',
                  '-': '||Dash||',
                  '\n': '||Return||'}   
    return token_dict
"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_tokenize(token_lookup)

Tests Passed


## Pre-process all the data and save it

Running the code cell below will pre-process all the data and save it to file. You're encouraged to lok at the code for `preprocess_and_save_data` in the `helpers.py` file to see what it's doing in detail, but you do not need to change this code.

In [47]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# pre-process training data
helper.preprocess_and_save_data(data_dir, token_lookup, create_lookup_tables)

# Check Point
This is your first checkpoint. If you ever decide to come back to this notebook or have to restart the notebook, you can start from here. The preprocessed data has been saved to disk.

In [1]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import helper
import problem_unittests as tests

int_text, vocab_to_int, int_to_vocab, token_dict = helper.load_preprocess()

## Build the Neural Network
In this section, you'll build the components necessary to build an RNN by implementing the RNN Module and forward and backpropagation functions.

### Check Access to GPU

In [2]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import torch

# Check for a GPU
train_on_gpu = torch.cuda.is_available()
if not train_on_gpu:
    print('No GPU found. Please use a GPU to train your neural network.')

## Input
Let's start with the preprocessed input data. We'll use [TensorDataset](http://pytorch.org/docs/master/data.html#torch.utils.data.TensorDataset) to provide a known format to our dataset; in combination with [DataLoader](http://pytorch.org/docs/master/data.html#torch.utils.data.DataLoader), it will handle batching, shuffling, and other dataset iteration functions.

You can create data with TensorDataset by passing in feature and target tensors. Then create a DataLoader as usual.
```
data = TensorDataset(feature_tensors, target_tensors)
data_loader = torch.utils.data.DataLoader(data, 
                                          batch_size=batch_size)
```

### Batching
Implement the `batch_data` function to batch `words` data into chunks of size `batch_size` using the `TensorDataset` and `DataLoader` classes.

>You can batch words using the DataLoader, but it will be up to you to create `feature_tensors` and `target_tensors` of the correct size and content for a given `sequence_length`.

For example, say we have these as input:
```
words = [1, 2, 3, 4, 5, 6, 7]
sequence_length = 4
```

Your first `feature_tensor` should contain the values:
```
[1, 2, 3, 4]
```
And the corresponding `target_tensor` should just be the next "word"/tokenized word value:
```
5
```
This should continue with the second `feature_tensor`, `target_tensor` being:
```
[2, 3, 4, 5]  # features
6             # target
```

In [3]:
from torch.utils.data import TensorDataset, DataLoader
import numpy as np

def batch_data(words, sequence_length, batch_size):
    """
    Batch the neural network data using DataLoader
    :param words: The word ids of the TV scripts
    :param sequence_length: The sequence length of each batch
    :param batch_size: The size of each batch; the number of sequences in a batch
    :return: DataLoader with batched data
    """
    # TODO: Implement function
    n_batches = len(words)//(sequence_length * batch_size)
    words = np.array(words[:(n_batches*sequence_length*batch_size)])
    

    features, targets = [], []
    for i in range(0, len(words), sequence_length):
        feature = words[i:i+sequence_length]
        target = np.zeros_like(feature)
        try:
            target = words[i+sequence_length]
        except:
            target = words[0]
        
        features.append(feature)
        targets.append(target)
        
    #convert to tensors
    feature_tensors, target_tensors = torch.from_numpy(np.array(features)), torch.from_numpy(np.array(targets))
    
    data = TensorDataset(feature_tensors, target_tensors)
    data_loader = torch.utils.data.DataLoader(data, 
                                          batch_size=batch_size, shuffle = True)
    # return a dataloader
    return data_loader

# there is no test for this function, but you are encouraged to create
# print statements and tests of your own


### Test your dataloader 

You'll have to modify this code to test a batching function, but it should look fairly similar.

Below, we're generating some test text data and defining a dataloader using the function you defined, above. Then, we are getting some sample batch of inputs `sample_x` and targets `sample_y` from our dataloader.

Your code should return something like the following (likely in a different order, if you shuffled your data):

```
torch.Size([10, 5])
tensor([[ 28,  29,  30,  31,  32],
        [ 21,  22,  23,  24,  25],
        [ 17,  18,  19,  20,  21],
        [ 34,  35,  36,  37,  38],
        [ 11,  12,  13,  14,  15],
        [ 23,  24,  25,  26,  27],
        [  6,   7,   8,   9,  10],
        [ 38,  39,  40,  41,  42],
        [ 25,  26,  27,  28,  29],
        [  7,   8,   9,  10,  11]])

torch.Size([10])
tensor([ 33,  26,  22,  39,  16,  28,  11,  43,  30,  12])
```

### Sizes
Your sample_x should be of size `(batch_size, sequence_length)` or (10, 5) in this case and sample_y should just have one dimension: batch_size (10). 

### Values

You should also notice that the targets, sample_y, are the *next* value in the ordered test_text data. So, for an input sequence `[ 28,  29,  30,  31,  32]` that ends with the value `32`, the corresponding output should be `33`.

In [4]:
# test dataloader

test_text = range(50)
t_loader = batch_data(test_text, sequence_length=5, batch_size=10)

data_iter = iter(t_loader)
sample_x, sample_y = data_iter.next()

print(sample_x.shape)
print(sample_x)
print()
print(sample_y.shape)
print(sample_y)

torch.Size([10, 5])
tensor([[ 20,  21,  22,  23,  24],
        [  0,   1,   2,   3,   4],
        [ 25,  26,  27,  28,  29],
        [ 30,  31,  32,  33,  34],
        [ 35,  36,  37,  38,  39],
        [ 15,  16,  17,  18,  19],
        [ 10,  11,  12,  13,  14],
        [ 40,  41,  42,  43,  44],
        [  5,   6,   7,   8,   9],
        [ 45,  46,  47,  48,  49]])

torch.Size([10])
tensor([ 25,   5,  30,  35,  40,  20,  15,  45,  10,   0])


---
## Build the Neural Network
Implement an RNN using PyTorch's [Module class](http://pytorch.org/docs/master/nn.html#torch.nn.Module). You may choose to use a GRU or an LSTM. To complete the RNN, you'll have to implement the following functions for the class:
 - `__init__` - The initialize function. 
 - `init_hidden` - The initialization function for an LSTM/GRU hidden state
 - `forward` - Forward propagation function.
 
The initialize function should create the layers of the neural network and save them to the class. The forward propagation function will use these layers to run forward propagation and generate an output and a hidden state.

**The output of this model should be the *last* batch of word scores** after a complete sequence has been processed. That is, for each input sequence of words, we only want to output the word scores for a single, most likely, next word.

### Hints

1. Make sure to stack the outputs of the lstm to pass to your fully-connected layer, you can do this with `lstm_output = lstm_output.contiguous().view(-1, self.hidden_dim)`
2. You can get the last batch of word scores by shaping the output of the final, fully-connected layer like so:

```
# reshape into (batch_size, seq_length, output_size)
output = output.view(batch_size, -1, self.output_size)
# get last batch
out = output[:, -1]
```

In [5]:
import torch.nn as nn

class RNN(nn.Module):
    
    def __init__(self, vocab_size, output_size, embedding_dim, hidden_dim, n_layers, dropout=0.5):
        """
        Initialize the PyTorch RNN Module
        :param vocab_size: The number of input dimensions of the neural network (the size of the vocabulary)
        :param output_size: The number of output dimensions of the neural network
        :param embedding_dim: The size of embeddings, should you choose to use them        
        :param hidden_dim: The size of the hidden layer outputs
        :param dropout: dropout to add in between LSTM/GRU layers
        """
        super(RNN, self).__init__()
        # TODO: Implement function
        
        # set class variables
        self.output_size = output_size
        self.hidden_dim = hidden_dim
        self.n_layers = n_layers
        # define model layers
        self.embed = nn.Embedding(vocab_size, embedding_dim)
        self.lstm = nn.LSTM(embedding_dim, self.hidden_dim, self.n_layers, batch_first=True, dropout = dropout)
        self.fc = nn.Linear(self.hidden_dim, self.output_size)
       
        self.dropout = nn.Dropout(p=dropout)
    
    
    def forward(self, nn_input, hidden):
        """
        Forward propagation of the neural network
        :param nn_input: The input to the neural network
        :param hidden: The hidden state        
        :return: Two Tensors, the output of the neural network and the latest hidden state
        """
        # TODO: Implement function  
        batch_size = nn_input.shape[0]
        embed_input = self.embed(nn_input)
        lstm_output, hidden = self.lstm(embed_input, hidden)
        lstm_output = lstm_output.contiguous().view(-1, self.hidden_dim)
        output = self.dropout(lstm_output)
        output = self.fc(output)
        
        output = output.view(batch_size, -1, self.output_size)
        out = output[:, -1]
        # return one batch of output word scores and the hidden state
        return out, hidden
    
    
    def init_hidden(self, batch_size):
        '''
        Initialize the hidden state of an LSTM/GRU
        :param batch_size: The batch_size of the hidden state
        :return: hidden state of dims (n_layers, batch_size, hidden_dim)
        '''
        # Implement function
        
        # initialize hidden state with zero weights, and move to GPU if available
        if (train_on_gpu):
            hidden = (next(self.parameters()).data.new(self.n_layers, batch_size, self.hidden_dim).zero_().cuda(),
                  next(self.parameters()).data.new(self.n_layers, batch_size, self.hidden_dim).zero_().cuda())
        else:
            hidden = (next(self.parameters()).data.new(self.n_layers, batch_size, self.hidden_dim).zero_(),
                      next(self.parameters()).data.new(self.n_layers, batch_size, self.hidden_dim).zero_())
        return hidden

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_rnn(RNN, train_on_gpu)

Tests Passed


### Define forward and backpropagation

Use the RNN class you implemented to apply forward and back propagation. This function will be called, iteratively, in the training loop as follows:
```
loss = forward_back_prop(decoder, decoder_optimizer, criterion, inp, target)
```

And it should return the average loss over a batch and the hidden state returned by a call to `RNN(inp, hidden)`. Recall that you can get this loss by computing it, as usual, and calling `loss.item()`.

**If a GPU is available, you should move your data to that GPU device, here.**

In [6]:
def forward_back_prop(rnn, optimizer, criterion, inp, target, hidden):
    """
    Forward and backward propagation on the neural network
    :param decoder: The PyTorch Module that holds the neural network
    :param decoder_optimizer: The PyTorch optimizer for the neural network
    :param criterion: The PyTorch loss function
    :param inp: A batch of input to the neural network
    :param target: The target output for the batch of input
    :return: The loss and the latest hidden state Tensor
    """
    
    # TODO: Implement Function
    
    # move data to GPU, if available
    train_on_gpu = torch.cuda.is_available()
    if (train_on_gpu):
           rnn, inp, target = rnn.cuda(), inp.cuda(), target.cuda()
    # perform backpropagation and optimization
    hidden = tuple([each.data for each in hidden])
    optimizer.zero_grad()
    output, hidden = rnn(inp, hidden)
    loss = criterion(output.squeeze(), target)
    loss.backward()
    optimizer.step()
    # return the loss over a batch and the hidden state produced by our model
    return loss.item(), hidden

# Note that these tests aren't completely extensive.
# they are here to act as general checks on the expected outputs of your functions
"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_forward_back_prop(RNN, forward_back_prop, train_on_gpu)

Tests Passed


## Neural Network Training

With the structure of the network complete and data ready to be fed in the neural network, it's time to train it.

### Train Loop

The training loop is implemented for you in the `train_decoder` function. This function will train the network over all the batches for the number of epochs given. The model progress will be shown every number of batches. This number is set with the `show_every_n_batches` parameter. You'll set this parameter along with other parameters in the next section.

In [7]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""

def train_rnn(rnn, batch_size, optimizer, criterion, n_epochs, show_every_n_batches=100):
    batch_losses = []
    
    rnn.train()

    print("Training for %d epoch(s)..." % n_epochs)
    for epoch_i in range(1, n_epochs + 1):
        
        # initialize hidden state
        hidden = rnn.init_hidden(batch_size)
        
        for batch_i, (inputs, labels) in enumerate(train_loader, 1):
            
            # make sure you iterate over completely full batches, only
            n_batches = len(train_loader.dataset)//batch_size
            if(batch_i > n_batches):
                break
            
            # forward, back prop
            loss, hidden = forward_back_prop(rnn, optimizer, criterion, inputs, labels, hidden)          
            # record loss
            batch_losses.append(loss)

            # printing loss stats
            if batch_i % show_every_n_batches == 0:
                print('Epoch: {:>4}/{:<4}  Loss: {}\n'.format(
                    epoch_i, n_epochs, np.average(batch_losses)))
                batch_losses = []

    # returns a trained rnn
    return rnn

### Hyperparameters

Set and train the neural network with the following parameters:
- Set `sequence_length` to the length of a sequence.
- Set `batch_size` to the batch size.
- Set `num_epochs` to the number of epochs to train for.
- Set `learning_rate` to the learning rate for an Adam optimizer.
- Set `vocab_size` to the number of uniqe tokens in our vocabulary.
- Set `output_size` to the desired size of the output.
- Set `embedding_dim` to the embedding dimension; smaller than the vocab_size.
- Set `hidden_dim` to the hidden dimension of your RNN.
- Set `n_layers` to the number of layers/cells in your RNN.
- Set `show_every_n_batches` to the number of batches at which the neural network should print progress.

If the network isn't getting the desired results, tweak these parameters and/or the layers in the `RNN` class.

In [8]:
print(len(int_text))

892110


In [9]:
# Data params
# Sequence Length
sequence_length = 6 # of words in a sequence
# Batch Size
batch_size = 64

# data loader - do not change
train_loader = batch_data(int_text, sequence_length, batch_size)

In [10]:
# Training parameters
# Number of Epochs
num_epochs = 30
# Learning Rate
learning_rate = 0.001

# Model parameters
# Vocab size
vocab_size = len(vocab_to_int) 
# Output size
output_size = vocab_size
# Embedding Dimension
embedding_dim = 500
# Hidden Dimension
hidden_dim = 512
# Number of RNN Layers
n_layers = 3

# Show stats for every n number of batches
show_every_n_batches = 10

### Train
In the next cell, you'll train the neural network on the pre-processed data.  If you have a hard time getting a good loss, you may consider changing your hyperparameters. In general, you may get better results with larger hidden and n_layer dimensions, but larger models take a longer time to train. 
> **You should aim for a loss less than 3.5.** 

You should also experiment with different sequence lengths, which determine the size of the long range dependencies that a model can learn.

In [ ]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""

# create model and move to gpu if available
rnn = RNN(vocab_size, output_size, embedding_dim, hidden_dim, n_layers, dropout=0.5)
if train_on_gpu:
    rnn.cuda()

# defining loss and optimization functions for training
optimizer = torch.optim.Adam(rnn.parameters(), lr=learning_rate)
criterion = nn.CrossEntropyLoss()

# training the model
trained_rnn = train_rnn(rnn, batch_size, optimizer, criterion, num_epochs, show_every_n_batches)

# saving the trained model
helper.save_model('./save/trained_rnn', trained_rnn)
print('Model Trained and Saved')

Training for 30 epoch(s)...
Epoch:    1/30    Loss: 9.084050607681274

Epoch:    1/30    Loss: 6.995461559295654

Epoch:    1/30    Loss: 6.652137517929077

Epoch:    1/30    Loss: 6.269296312332154

Epoch:    1/30    Loss: 6.156131982803345

Epoch:    1/30    Loss: 5.983782958984375

Epoch:    1/30    Loss: 6.086099529266358

Epoch:    1/30    Loss: 5.864189195632934

Epoch:    1/30    Loss: 6.206787014007569

Epoch:    1/30    Loss: 5.815178251266479

Epoch:    1/30    Loss: 5.874624156951905

Epoch:    1/30    Loss: 5.590088748931885

Epoch:    1/30    Loss: 5.80139012336731

Epoch:    1/30    Loss: 5.8248138427734375

Epoch:    1/30    Loss: 5.81760950088501

Epoch:    1/30    Loss: 5.4485344886779785

Epoch:    1/30    Loss: 5.7600555419921875

Epoch:    1/30    Loss: 5.5437352657318115

Epoch:    1/30    Loss: 5.3279420852661135

Epoch:    1/30    Loss: 5.54026951789856

Epoch:    1/30    Loss: 5.70044322013855

Epoch:    1/30    Loss: 5.886346292495728

Epoch:    1/30    Loss: 5

Epoch:    1/30    Loss: 4.890138721466064

Epoch:    1/30    Loss: 4.910463953018189

Epoch:    1/30    Loss: 4.929575443267822

Epoch:    1/30    Loss: 4.96461820602417

Epoch:    1/30    Loss: 4.78700635433197

Epoch:    1/30    Loss: 4.730292940139771

Epoch:    1/30    Loss: 4.921347379684448

Epoch:    1/30    Loss: 4.662320280075074

Epoch:    1/30    Loss: 5.032066774368286

Epoch:    1/30    Loss: 4.687074756622314

Epoch:    1/30    Loss: 4.856114435195923

Epoch:    1/30    Loss: 4.982973861694336

Epoch:    1/30    Loss: 5.052369165420532

Epoch:    1/30    Loss: 4.748007106781006

Epoch:    1/30    Loss: 4.7107847213745115

Epoch:    1/30    Loss: 4.678832483291626

Epoch:    1/30    Loss: 4.732467079162598

Epoch:    1/30    Loss: 4.891993236541748

Epoch:    1/30    Loss: 4.686230421066284

Epoch:    1/30    Loss: 4.672486734390259

Epoch:    1/30    Loss: 4.765953254699707

Epoch:    1/30    Loss: 4.784014344215393

Epoch:    1/30    Loss: 4.756644773483276

Epoch:    1/

Epoch:    2/30    Loss: 4.691123199462891

Epoch:    2/30    Loss: 4.601697683334351

Epoch:    2/30    Loss: 4.431751251220703

Epoch:    2/30    Loss: 4.697372722625732

Epoch:    2/30    Loss: 4.517445707321167

Epoch:    2/30    Loss: 4.639889240264893

Epoch:    2/30    Loss: 4.808301877975464

Epoch:    2/30    Loss: 4.339040517807007

Epoch:    2/30    Loss: 4.553295421600342

Epoch:    2/30    Loss: 4.63047513961792

Epoch:    2/30    Loss: 4.205301094055176

Epoch:    2/30    Loss: 4.499992847442627

Epoch:    2/30    Loss: 4.658674192428589

Epoch:    2/30    Loss: 4.565420341491699

Epoch:    2/30    Loss: 4.403257036209107

Epoch:    2/30    Loss: 4.481516170501709

Epoch:    2/30    Loss: 4.500877141952515

Epoch:    2/30    Loss: 4.396097898483276

Epoch:    2/30    Loss: 4.715815782546997

Epoch:    2/30    Loss: 4.702153873443604

Epoch:    2/30    Loss: 4.443569231033325

Epoch:    2/30    Loss: 4.670051097869873

Epoch:    2/30    Loss: 4.445018315315247

Epoch:    2/

Epoch:    3/30    Loss: 4.697815489768982

Epoch:    3/30    Loss: 4.219480323791504

Epoch:    3/30    Loss: 4.301202321052552

Epoch:    3/30    Loss: 4.594737720489502

Epoch:    3/30    Loss: 4.379376292228699

Epoch:    3/30    Loss: 4.279119277000428

Epoch:    3/30    Loss: 4.224784970283508

Epoch:    3/30    Loss: 4.409577822685241

Epoch:    3/30    Loss: 4.18730411529541

Epoch:    3/30    Loss: 4.5269163131713865

Epoch:    3/30    Loss: 4.393619132041931

Epoch:    3/30    Loss: 4.233287763595581

Epoch:    3/30    Loss: 4.717074060440064

Epoch:    3/30    Loss: 4.34702570438385

Epoch:    3/30    Loss: 4.575814747810364

Epoch:    3/30    Loss: 4.349825143814087

Epoch:    3/30    Loss: 4.25186333656311

Epoch:    3/30    Loss: 4.362956476211548

Epoch:    3/30    Loss: 4.4029250860214235

Epoch:    3/30    Loss: 4.547167086601258

Epoch:    3/30    Loss: 4.417986583709717

Epoch:    3/30    Loss: 4.2847585439682

Epoch:    3/30    Loss: 4.435477900505066

Epoch:    3/30

Epoch:    4/30    Loss: 4.160111308097839

Epoch:    4/30    Loss: 4.290582299232483

Epoch:    4/30    Loss: 4.255142331123352

Epoch:    4/30    Loss: 4.172678518295288

Epoch:    4/30    Loss: 4.323028564453125

Epoch:    4/30    Loss: 4.428517389297485

Epoch:    4/30    Loss: 4.256661725044251

Epoch:    4/30    Loss: 4.314753484725952

Epoch:    4/30    Loss: 4.505028963088989

Epoch:    4/30    Loss: 4.452885627746582

Epoch:    4/30    Loss: 4.188143610954285

Epoch:    4/30    Loss: 4.239332365989685

Epoch:    4/30    Loss: 4.129481339454651

Epoch:    4/30    Loss: 4.082048678398133

Epoch:    4/30    Loss: 4.215937399864197

Epoch:    4/30    Loss: 4.107482075691223

Epoch:    4/30    Loss: 4.250562381744385

Epoch:    4/30    Loss: 4.299384951591492

Epoch:    4/30    Loss: 4.100442719459534

Epoch:    4/30    Loss: 4.133211636543274

Epoch:    4/30    Loss: 4.255440759658813

Epoch:    4/30    Loss: 4.345036745071411

Epoch:    4/30    Loss: 4.539964103698731

Epoch:    4

Epoch:    5/30    Loss: 4.10638575553894

Epoch:    5/30    Loss: 4.167181611061096

Epoch:    5/30    Loss: 4.171488428115845

Epoch:    5/30    Loss: 4.193336725234985

Epoch:    5/30    Loss: 4.110875105857849

Epoch:    5/30    Loss: 3.9567234754562377

Epoch:    5/30    Loss: 4.223464298248291

Epoch:    5/30    Loss: 4.316490721702576

Epoch:    5/30    Loss: 4.027927279472351

Epoch:    5/30    Loss: 4.213637638092041

Epoch:    5/30    Loss: 4.099527096748352

Epoch:    5/30    Loss: 4.031114315986633

Epoch:    5/30    Loss: 3.926592254638672

Epoch:    5/30    Loss: 4.185042786598205

Epoch:    5/30    Loss: 4.420453214645386

Epoch:    5/30    Loss: 3.876425552368164

Epoch:    5/30    Loss: 4.155665588378906

Epoch:    5/30    Loss: 4.252432775497437

Epoch:    5/30    Loss: 3.9774825096130373

Epoch:    5/30    Loss: 4.118820595741272

Epoch:    5/30    Loss: 4.029390835762024

Epoch:    5/30    Loss: 4.183443880081176

Epoch:    5/30    Loss: 4.339428997039795

Epoch:    

Epoch:    5/30    Loss: 3.9193825721740723

Epoch:    5/30    Loss: 4.094112372398376

Epoch:    5/30    Loss: 4.191403317451477

Epoch:    5/30    Loss: 4.109126687049866

Epoch:    5/30    Loss: 4.146739077568054

Epoch:    5/30    Loss: 3.977791261672974

Epoch:    5/30    Loss: 4.2362720489501955

Epoch:    5/30    Loss: 4.072604131698609

Epoch:    5/30    Loss: 3.99286835193634

Epoch:    5/30    Loss: 4.0934504747390745

Epoch:    5/30    Loss: 4.284762883186341

Epoch:    5/30    Loss: 4.055362391471863

Epoch:    5/30    Loss: 4.151039886474609

Epoch:    5/30    Loss: 4.068847870826721

Epoch:    5/30    Loss: 4.129468250274658

Epoch:    6/30    Loss: 3.8350011752201962

Epoch:    6/30    Loss: 4.036361885070801

Epoch:    6/30    Loss: 4.001075410842896

Epoch:    6/30    Loss: 4.000904941558838

Epoch:    6/30    Loss: 4.161778211593628

Epoch:    6/30    Loss: 4.001233506202698

Epoch:    6/30    Loss: 4.00858793258667

Epoch:    6/30    Loss: 3.880178952217102

Epoch:   

Epoch:    6/30    Loss: 4.16809983253479

Epoch:    6/30    Loss: 4.045735621452332

Epoch:    6/30    Loss: 3.815611553192139

Epoch:    6/30    Loss: 3.9668652772903443

Epoch:    6/30    Loss: 3.932171082496643

Epoch:    6/30    Loss: 4.102631092071533

Epoch:    6/30    Loss: 4.214882969856262

Epoch:    6/30    Loss: 4.203335309028626

Epoch:    6/30    Loss: 4.036382102966309

Epoch:    6/30    Loss: 4.057118773460388

Epoch:    6/30    Loss: 4.185410690307617

Epoch:    6/30    Loss: 3.92538583278656

Epoch:    6/30    Loss: 3.8717113733291626

Epoch:    6/30    Loss: 4.093151259422302

Epoch:    6/30    Loss: 4.232369709014892

Epoch:    6/30    Loss: 4.116027188301087

Epoch:    6/30    Loss: 4.14262626171112

Epoch:    6/30    Loss: 4.174790573120117

Epoch:    6/30    Loss: 3.9417344093322755

Epoch:    6/30    Loss: 4.107788753509522

Epoch:    6/30    Loss: 4.140357446670532

Epoch:    6/30    Loss: 4.015265369415284

Epoch:    6/30    Loss: 3.9506715536117554

Epoch:    

Epoch:    7/30    Loss: 4.092762780189514

Epoch:    7/30    Loss: 3.743574023246765

Epoch:    7/30    Loss: 4.039117622375488

Epoch:    7/30    Loss: 3.9713380336761475

Epoch:    7/30    Loss: 4.064222359657288

Epoch:    7/30    Loss: 4.041673159599304

Epoch:    7/30    Loss: 4.2313416481018065

Epoch:    7/30    Loss: 3.9963254928588867

Epoch:    7/30    Loss: 3.993784213066101

Epoch:    7/30    Loss: 3.8185434341430664

Epoch:    7/30    Loss: 4.049813294410706

Epoch:    7/30    Loss: 3.8798056125640867

Epoch:    7/30    Loss: 3.901490092277527

Epoch:    7/30    Loss: 3.9236759662628176

Epoch:    7/30    Loss: 3.895812678337097

Epoch:    7/30    Loss: 3.903115439414978

Epoch:    7/30    Loss: 3.6818843841552735

Epoch:    7/30    Loss: 3.9406662225723266

Epoch:    7/30    Loss: 3.8917083501815797

Epoch:    7/30    Loss: 3.929228734970093

Epoch:    7/30    Loss: 3.8541561365127563

Epoch:    7/30    Loss: 4.033877992630005

Epoch:    7/30    Loss: 3.8896217346191406



Epoch:    8/30    Loss: 3.880265808105469

Epoch:    8/30    Loss: 3.789081835746765

Epoch:    8/30    Loss: 3.8847885131835938

Epoch:    8/30    Loss: 3.6041170597076415

Epoch:    8/30    Loss: 3.859546589851379

Epoch:    8/30    Loss: 3.7905864477157594

Epoch:    8/30    Loss: 3.96161425113678

Epoch:    8/30    Loss: 3.9150072813034056

Epoch:    8/30    Loss: 3.6987080812454223

Epoch:    8/30    Loss: 4.152738547325134

Epoch:    8/30    Loss: 3.927053713798523

Epoch:    8/30    Loss: 3.810522723197937

Epoch:    8/30    Loss: 3.806903028488159

Epoch:    8/30    Loss: 3.947902226448059

Epoch:    8/30    Loss: 3.737462639808655

Epoch:    8/30    Loss: 4.05812017917633

Epoch:    8/30    Loss: 3.8301543474197386

Epoch:    8/30    Loss: 3.863547110557556

Epoch:    8/30    Loss: 3.975518560409546

Epoch:    8/30    Loss: 3.8932446956634523

Epoch:    8/30    Loss: 3.8905158042907715

Epoch:    8/30    Loss: 3.9591373682022093

Epoch:    8/30    Loss: 3.8474576234817506

Epo

Epoch:    9/30    Loss: 3.7390584468841555

Epoch:    9/30    Loss: 3.73633816242218

Epoch:    9/30    Loss: 3.904834342002869

Epoch:    9/30    Loss: 3.787322163581848

Epoch:    9/30    Loss: 3.883016800880432

Epoch:    9/30    Loss: 3.8053159713745117

Epoch:    9/30    Loss: 3.8982678413391114

Epoch:    9/30    Loss: 3.902544045448303

Epoch:    9/30    Loss: 3.8029676914215087

Epoch:    9/30    Loss: 3.614639639854431

Epoch:    9/30    Loss: 3.6826927423477174

Epoch:    9/30    Loss: 3.9780714750289916

Epoch:    9/30    Loss: 3.732434344291687

Epoch:    9/30    Loss: 3.8131590127944945

Epoch:    9/30    Loss: 3.693909788131714

Epoch:    9/30    Loss: 3.8428542852401733

Epoch:    9/30    Loss: 3.538834977149963

Epoch:    9/30    Loss: 3.961109232902527

Epoch:    9/30    Loss: 3.79898054599762

Epoch:    9/30    Loss: 3.816526699066162

Epoch:    9/30    Loss: 3.7843088865280152

Epoch:    9/30    Loss: 3.533921790122986

Epoch:    9/30    Loss: 3.6708391189575194

Epo

Epoch:   10/30    Loss: 3.6462326526641844

Epoch:   10/30    Loss: 3.6584047794342043

Epoch:   10/30    Loss: 3.663607430458069

Epoch:   10/30    Loss: 3.743320894241333

Epoch:   10/30    Loss: 3.81061429977417

Epoch:   10/30    Loss: 3.604158353805542

Epoch:   10/30    Loss: 3.395508313179016

Epoch:   10/30    Loss: 3.5712238550186157

Epoch:   10/30    Loss: 3.597188210487366

Epoch:   10/30    Loss: 3.6616553544998167

Epoch:   10/30    Loss: 3.682467484474182

Epoch:   10/30    Loss: 3.757941246032715

Epoch:   10/30    Loss: 3.6958032608032227

Epoch:   10/30    Loss: 3.7580267667770384

Epoch:   10/30    Loss: 3.7735061645507812

Epoch:   10/30    Loss: 3.829426980018616

Epoch:   10/30    Loss: 3.543812561035156

Epoch:   10/30    Loss: 3.608244276046753

Epoch:   10/30    Loss: 3.5272532224655153

Epoch:   10/30    Loss: 3.548163962364197

Epoch:   10/30    Loss: 3.707163119316101

Epoch:   10/30    Loss: 3.733958697319031

Epoch:   10/30    Loss: 3.685978889465332

Epoc

Epoch:   10/30    Loss: 3.8652780771255495

Epoch:   10/30    Loss: 3.669055700302124

Epoch:   10/30    Loss: 3.810415196418762

Epoch:   10/30    Loss: 3.7825940608978272

Epoch:   10/30    Loss: 3.7463526487350465

Epoch:   10/30    Loss: 3.7734228372573853

Epoch:   10/30    Loss: 3.9338727235794066

Epoch:   10/30    Loss: 3.6804303407669066

Epoch:   10/30    Loss: 3.9332821369171143

Epoch:   10/30    Loss: 3.6732664346694945

Epoch:   10/30    Loss: 3.807285737991333

Epoch:   10/30    Loss: 3.577465629577637

Epoch:   10/30    Loss: 4.05613763332367

Epoch:   10/30    Loss: 3.5270774126052857

Epoch:   10/30    Loss: 4.096084094047546

Epoch:   10/30    Loss: 3.7300278902053834

Epoch:   10/30    Loss: 3.842986512184143

Epoch:   10/30    Loss: 3.873705840110779

Epoch:   10/30    Loss: 3.869455122947693

Epoch:   10/30    Loss: 3.797688865661621

Epoch:   10/30    Loss: 3.781400513648987

Epoch:   10/30    Loss: 3.8227349281311036

Epoch:   10/30    Loss: 3.773143839836121

E

Epoch:   11/30    Loss: 3.57256178855896

Epoch:   11/30    Loss: 3.639573264122009

Epoch:   11/30    Loss: 3.631300449371338

Epoch:   11/30    Loss: 3.7543514251708983

Epoch:   11/30    Loss: 3.5068052053451537

Epoch:   11/30    Loss: 3.6827163219451906

Epoch:   11/30    Loss: 3.6401017189025877

Epoch:   11/30    Loss: 3.8448739051818848

Epoch:   11/30    Loss: 3.674321508407593

Epoch:   11/30    Loss: 3.633661127090454

Epoch:   11/30    Loss: 3.8032937049865723

Epoch:   11/30    Loss: 3.6621679067611694

Epoch:   11/30    Loss: 3.761248755455017

Epoch:   11/30    Loss: 3.6443475246429444

Epoch:   11/30    Loss: 3.6681393146514893

Epoch:   11/30    Loss: 3.820710778236389

Epoch:   11/30    Loss: 3.6062906265258787

Epoch:   11/30    Loss: 3.792156529426575

Epoch:   11/30    Loss: 3.769367527961731

Epoch:   11/30    Loss: 3.524245023727417

Epoch:   11/30    Loss: 3.8332365274429323

Epoch:   11/30    Loss: 3.5214528799057008

Epoch:   11/30    Loss: 3.619246983528137



Epoch:   12/30    Loss: 3.6045965909957887

Epoch:   12/30    Loss: 3.6065662622451784

Epoch:   12/30    Loss: 3.4044270515441895

Epoch:   12/30    Loss: 3.7138018369674684

Epoch:   12/30    Loss: 3.5975170373916625

Epoch:   12/30    Loss: 3.4626824617385865

Epoch:   12/30    Loss: 3.6664140939712526

Epoch:   12/30    Loss: 3.6037444353103636

Epoch:   12/30    Loss: 3.4657127618789674

Epoch:   12/30    Loss: 3.6881954431533814

Epoch:   12/30    Loss: 3.5285595655441284

Epoch:   12/30    Loss: 3.6917500734329223

Epoch:   12/30    Loss: 3.5908459424972534

Epoch:   12/30    Loss: 3.6064014196395875

Epoch:   12/30    Loss: 3.4826735973358156

Epoch:   12/30    Loss: 3.577062201499939

Epoch:   12/30    Loss: 3.8230380058288573

Epoch:   12/30    Loss: 3.7887808799743654

Epoch:   12/30    Loss: 3.4243078231811523

Epoch:   12/30    Loss: 3.667731833457947

Epoch:   12/30    Loss: 3.5566087484359743

Epoch:   12/30    Loss: 3.611651134490967

Epoch:   12/30    Loss: 3.541546511

Epoch:   13/30    Loss: 3.6436277627944946

Epoch:   13/30    Loss: 3.6040177583694457

Epoch:   13/30    Loss: 3.4628187656402587

Epoch:   13/30    Loss: 3.5390259265899657

Epoch:   13/30    Loss: 3.545737552642822

Epoch:   13/30    Loss: 3.3219829320907595

Epoch:   13/30    Loss: 3.5745007514953615

Epoch:   13/30    Loss: 3.543532967567444

Epoch:   13/30    Loss: 3.8067395210266115

Epoch:   13/30    Loss: 3.603104662895203

Epoch:   13/30    Loss: 3.352462315559387

Epoch:   13/30    Loss: 3.600733423233032

Epoch:   13/30    Loss: 3.587279224395752

Epoch:   13/30    Loss: 3.385334277153015

Epoch:   13/30    Loss: 3.573415160179138

Epoch:   13/30    Loss: 3.5727209091186523

Epoch:   13/30    Loss: 3.495465135574341

Epoch:   13/30    Loss: 3.574917769432068

Epoch:   13/30    Loss: 3.3432307481765746

Epoch:   13/30    Loss: 3.6973682403564454

Epoch:   13/30    Loss: 3.641809415817261

Epoch:   13/30    Loss: 3.6066526651382445

Epoch:   13/30    Loss: 3.6295364141464233


Epoch:   14/30    Loss: 3.3970542430877684

Epoch:   14/30    Loss: 3.6277410268783568

Epoch:   14/30    Loss: 3.4761767625808715

Epoch:   14/30    Loss: 3.354251837730408

Epoch:   14/30    Loss: 3.35067138671875

Epoch:   14/30    Loss: 3.4682471990585326

Epoch:   14/30    Loss: 3.6797577857971193

Epoch:   14/30    Loss: 3.4153606414794924

Epoch:   14/30    Loss: 3.3631651401519775

Epoch:   14/30    Loss: 3.4520944118499757

Epoch:   14/30    Loss: 3.4167787790298463

Epoch:   14/30    Loss: 3.494210362434387

Epoch:   14/30    Loss: 3.35142982006073

Epoch:   14/30    Loss: 3.3229782342910767

Epoch:   14/30    Loss: 3.4917705774307253

Epoch:   14/30    Loss: 3.394460129737854

Epoch:   14/30    Loss: 3.233443593978882

Epoch:   14/30    Loss: 3.5048582792282104

Epoch:   14/30    Loss: 3.4751150608062744

Epoch:   14/30    Loss: 3.376385140419006

Epoch:   14/30    Loss: 3.4303177118301393

Epoch:   14/30    Loss: 3.3696831703186034

Epoch:   14/30    Loss: 3.596552968025207

Epoch:   14/30    Loss: 3.5189669132232666

Epoch:   14/30    Loss: 3.6089792490005492

Epoch:   14/30    Loss: 3.589373540878296

Epoch:   14/30    Loss: 3.6377612113952638

Epoch:   14/30    Loss: 3.589463448524475

Epoch:   14/30    Loss: 3.6251749753952027

Epoch:   14/30    Loss: 3.6280375957489013

Epoch:   14/30    Loss: 3.4640448570251463

Epoch:   14/30    Loss: 3.62108519077301

Epoch:   14/30    Loss: 3.5542587518692015

Epoch:   14/30    Loss: 3.4245330572128294

Epoch:   14/30    Loss: 3.6090887546539308

Epoch:   14/30    Loss: 3.489687371253967

Epoch:   14/30    Loss: 3.4655433177947996

Epoch:   14/30    Loss: 3.4785784244537354

Epoch:   14/30    Loss: 3.621560215950012

Epoch:   14/30    Loss: 3.457602787017822

Epoch:   14/30    Loss: 3.595522952079773

Epoch:   14/30    Loss: 3.6294259786605836

Epoch:   14/30    Loss: 3.6979954481124877

Epoch:   14/30    Loss: 3.7193757772445677

Epoch:   14/30    Loss: 3.434201741218567

Epoch:   14/30    Loss: 3.563801503181457

Epoch:   15/30    Loss: 3.5299456119537354

Epoch:   15/30    Loss: 3.452171277999878

Epoch:   15/30    Loss: 3.3565082788467406

Epoch:   15/30    Loss: 3.4453275918960573

Epoch:   15/30    Loss: 3.5818046808242796

Epoch:   15/30    Loss: 3.4943567752838134

Epoch:   15/30    Loss: 3.4841639518737795

Epoch:   15/30    Loss: 3.685930061340332

Epoch:   15/30    Loss: 3.6212942361831666

Epoch:   15/30    Loss: 3.6308119535446166

Epoch:   15/30    Loss: 3.5910120010375977

Epoch:   15/30    Loss: 3.36774582862854

Epoch:   15/30    Loss: 3.3984490394592286

Epoch:   15/30    Loss: 3.5332184076309203

Epoch:   15/30    Loss: 3.449000096321106

Epoch:   15/30    Loss: 3.3860283136367797

Epoch:   15/30    Loss: 3.2336236476898192

Epoch:   15/30    Loss: 3.5143200397491454

Epoch:   15/30    Loss: 3.601057434082031

Epoch:   15/30    Loss: 3.650630235671997

Epoch:   15/30    Loss: 3.609057664871216

Epoch:   15/30    Loss: 3.3899625301361085

Epoch:   15/30    Loss: 3.50508360862731

Epoch:   16/30    Loss: 3.4554576873779297

Epoch:   16/30    Loss: 3.597635340690613

Epoch:   16/30    Loss: 3.384328103065491

Epoch:   16/30    Loss: 3.4305087327957153

Epoch:   16/30    Loss: 3.308612823486328

Epoch:   16/30    Loss: 3.605933737754822

Epoch:   16/30    Loss: 3.1383253574371337

Epoch:   16/30    Loss: 3.4692787885665894

Epoch:   16/30    Loss: 3.505332684516907

Epoch:   16/30    Loss: 3.512420988082886

Epoch:   16/30    Loss: 3.407169723510742

Epoch:   16/30    Loss: 3.6107214212417604

Epoch:   16/30    Loss: 3.4136035442352295

Epoch:   16/30    Loss: 3.461288332939148

Epoch:   16/30    Loss: 3.3624767303466796

Epoch:   16/30    Loss: 3.382170557975769

Epoch:   16/30    Loss: 3.3870654821395876

Epoch:   16/30    Loss: 3.6883471488952635

Epoch:   16/30    Loss: 3.429653263092041

Epoch:   16/30    Loss: 3.416927218437195

Epoch:   16/30    Loss: 3.477044939994812

Epoch:   16/30    Loss: 3.442146968841553

Epoch:   16/30    Loss: 3.3713456630706786

E

Epoch:   17/30    Loss: 3.394830107688904

Epoch:   17/30    Loss: 3.2134228229522703

Epoch:   17/30    Loss: 3.4091223001480104

Epoch:   17/30    Loss: 3.3630169868469237

Epoch:   17/30    Loss: 3.3445831537246704

Epoch:   17/30    Loss: 3.310818910598755

Epoch:   17/30    Loss: 3.403298044204712

Epoch:   17/30    Loss: 3.186805820465088

Epoch:   17/30    Loss: 3.2948216676712034

Epoch:   17/30    Loss: 3.1498391151428224

Epoch:   17/30    Loss: 3.4180026769638063

Epoch:   17/30    Loss: 3.346692657470703

Epoch:   17/30    Loss: 3.386559796333313

Epoch:   17/30    Loss: 3.4106470346450806

Epoch:   17/30    Loss: 3.5036712884902954

Epoch:   17/30    Loss: 3.325093936920166

Epoch:   17/30    Loss: 3.2368240356445312

Epoch:   17/30    Loss: 3.3158011436462402

Epoch:   17/30    Loss: 3.4118761301040648

Epoch:   17/30    Loss: 3.572712779045105

Epoch:   17/30    Loss: 3.258705568313599

Epoch:   17/30    Loss: 3.351902794837952

Epoch:   17/30    Loss: 3.4495944023132323

Epoch:   18/30    Loss: 3.1249476432800294

Epoch:   18/30    Loss: 3.323672389984131

Epoch:   18/30    Loss: 3.2456480741500853

Epoch:   18/30    Loss: 3.29434130191803

Epoch:   18/30    Loss: 3.293940544128418

Epoch:   18/30    Loss: 3.3417221546173095

Epoch:   18/30    Loss: 3.278832292556763

Epoch:   18/30    Loss: 3.488084650039673

Epoch:   18/30    Loss: 3.2893498659133913

Epoch:   18/30    Loss: 3.3113364696502687

Epoch:   18/30    Loss: 3.2531097650527956

Epoch:   18/30    Loss: 3.3414357900619507

Epoch:   18/30    Loss: 3.365761089324951

Epoch:   18/30    Loss: 3.243486833572388

Epoch:   18/30    Loss: 3.436632585525513

Epoch:   18/30    Loss: 3.2861133575439454

Epoch:   18/30    Loss: 3.151776075363159

Epoch:   18/30    Loss: 3.153056764602661

Epoch:   18/30    Loss: 3.4632569551467896

Epoch:   18/30    Loss: 3.1648911952972414

Epoch:   18/30    Loss: 3.2378511667251586

Epoch:   18/30    Loss: 3.5997130632400514

Epoch:   18/30    Loss: 3.2882911443710325


Epoch:   18/30    Loss: 3.387874126434326

Epoch:   18/30    Loss: 3.492685413360596

Epoch:   18/30    Loss: 3.31712646484375

Epoch:   18/30    Loss: 3.2913762807846068

Epoch:   18/30    Loss: 3.2810588836669923

Epoch:   18/30    Loss: 3.249550724029541

Epoch:   19/30    Loss: 3.270840259698721

Epoch:   19/30    Loss: 3.164074945449829

Epoch:   19/30    Loss: 3.3582830905914305

Epoch:   19/30    Loss: 3.3008846998214723

Epoch:   19/30    Loss: 3.1966615438461305

Epoch:   19/30    Loss: 3.2134724617004395

Epoch:   19/30    Loss: 3.127586245536804

Epoch:   19/30    Loss: 3.19575514793396

Epoch:   19/30    Loss: 3.2635095834732057

Epoch:   19/30    Loss: 3.262804126739502

Epoch:   19/30    Loss: 3.40944242477417

Epoch:   19/30    Loss: 3.3658448696136474

Epoch:   19/30    Loss: 3.2394179344177245

Epoch:   19/30    Loss: 3.2070631265640257

Epoch:   19/30    Loss: 3.39941246509552

Epoch:   19/30    Loss: 3.3005107402801515

Epoch:   19/30    Loss: 3.2795697450637817

Epo

Epoch:   19/30    Loss: 3.4012829303741454

Epoch:   19/30    Loss: 3.4890635013580322

Epoch:   19/30    Loss: 3.2667418241500856

Epoch:   19/30    Loss: 3.232177257537842

Epoch:   19/30    Loss: 3.3832767248153686

Epoch:   19/30    Loss: 3.4475083112716676

Epoch:   19/30    Loss: 3.2781127452850343

Epoch:   19/30    Loss: 3.3128586769104005

Epoch:   19/30    Loss: 3.387483739852905

Epoch:   19/30    Loss: 3.3763284921646117

Epoch:   19/30    Loss: 3.2630011081695556

Epoch:   19/30    Loss: 3.4040687322616576

Epoch:   19/30    Loss: 3.39250853061676

Epoch:   19/30    Loss: 3.499705457687378

Epoch:   19/30    Loss: 3.4250200986862183

Epoch:   19/30    Loss: 3.3890331983566284

Epoch:   19/30    Loss: 3.2751996994018553

Epoch:   19/30    Loss: 3.5031424283981325

Epoch:   19/30    Loss: 3.341634917259216

Epoch:   19/30    Loss: 3.457859826087952

Epoch:   19/30    Loss: 3.3450448274612428

Epoch:   19/30    Loss: 3.5773765802383424

Epoch:   19/30    Loss: 3.3556683778762

Epoch:   20/30    Loss: 3.401638221740723

Epoch:   20/30    Loss: 3.236457633972168

Epoch:   20/30    Loss: 3.367623543739319

Epoch:   20/30    Loss: 3.3534690141677856

Epoch:   20/30    Loss: 3.2074703931808473

Epoch:   20/30    Loss: 3.2035248994827272

Epoch:   20/30    Loss: 3.211257982254028

Epoch:   20/30    Loss: 3.145712208747864

Epoch:   20/30    Loss: 3.471935844421387

Epoch:   20/30    Loss: 3.231754016876221

Epoch:   20/30    Loss: 3.357793092727661

Epoch:   20/30    Loss: 3.2441171646118163

Epoch:   20/30    Loss: 3.288420057296753

Epoch:   20/30    Loss: 3.540961813926697

Epoch:   20/30    Loss: 3.4432471990585327

Epoch:   20/30    Loss: 3.3650779485702516

Epoch:   20/30    Loss: 3.3143808603286744

Epoch:   20/30    Loss: 3.225373888015747

Epoch:   20/30    Loss: 3.1848739624023437

Epoch:   20/30    Loss: 3.2930224418640135

Epoch:   20/30    Loss: 3.277019238471985

Epoch:   20/30    Loss: 3.3632161617279053

Epoch:   20/30    Loss: 3.446439814567566

E

Epoch:   21/30    Loss: 3.4276889324188233

Epoch:   21/30    Loss: 3.3828233242034913

Epoch:   21/30    Loss: 3.1524796724319457

Epoch:   21/30    Loss: 3.1320807456970217

Epoch:   21/30    Loss: 3.23438720703125

Epoch:   21/30    Loss: 3.248290705680847

Epoch:   21/30    Loss: 3.1995205879211426

Epoch:   21/30    Loss: 3.303974962234497

Epoch:   21/30    Loss: 3.1929324865341187

Epoch:   21/30    Loss: 3.078891158103943

Epoch:   21/30    Loss: 3.3154876232147217

Epoch:   21/30    Loss: 3.213479733467102

Epoch:   21/30    Loss: 2.976255679130554

Epoch:   21/30    Loss: 3.2927838563919067

Epoch:   21/30    Loss: 3.21415913105011

Epoch:   21/30    Loss: 3.1712615489959717

Epoch:   21/30    Loss: 3.2184029817581177

Epoch:   21/30    Loss: 3.284574270248413

Epoch:   21/30    Loss: 3.4066240072250364

Epoch:   21/30    Loss: 3.1815574169158936

Epoch:   21/30    Loss: 3.2523357629776

Epoch:   21/30    Loss: 3.333655261993408

Epoch:   21/30    Loss: 3.395615077018738

Epo

Epoch:   22/30    Loss: 3.1230700492858885

Epoch:   22/30    Loss: 3.2366558790206907

Epoch:   22/30    Loss: 3.160816502571106

Epoch:   22/30    Loss: 3.440940260887146

Epoch:   22/30    Loss: 3.2676218032836912

Epoch:   22/30    Loss: 3.008679485321045

Epoch:   22/30    Loss: 3.3383859157562257

Epoch:   22/30    Loss: 3.260417914390564

Epoch:   22/30    Loss: 3.129561257362366

Epoch:   22/30    Loss: 3.169590783119202

Epoch:   22/30    Loss: 3.080595588684082

Epoch:   22/30    Loss: 3.2319236516952516

Epoch:   22/30    Loss: 3.2377702951431275

Epoch:   22/30    Loss: 3.060767936706543

Epoch:   22/30    Loss: 3.111892080307007

Epoch:   22/30    Loss: 3.236559200286865

Epoch:   22/30    Loss: 3.0932528018951415

Epoch:   22/30    Loss: 3.1465978622436523

Epoch:   22/30    Loss: 3.224416995048523

Epoch:   22/30    Loss: 3.2019285202026366

Epoch:   22/30    Loss: 3.1447818279266357

Epoch:   22/30    Loss: 3.1412222146987916

Epoch:   22/30    Loss: 3.299162411689758



Epoch:   23/30    Loss: 2.9543654680252076

Epoch:   23/30    Loss: 3.0612967014312744

Epoch:   23/30    Loss: 3.3199177503585817

Epoch:   23/30    Loss: 3.1820005416870116

Epoch:   23/30    Loss: 3.1420325756073

Epoch:   23/30    Loss: 3.2400357007980345

Epoch:   23/30    Loss: 3.1701319217681885

Epoch:   23/30    Loss: 3.0982603073120116

Epoch:   23/30    Loss: 3.3346981763839723

Epoch:   23/30    Loss: 2.9457332491874695

Epoch:   23/30    Loss: 3.2414873838424683

Epoch:   23/30    Loss: 3.1612799882888796

Epoch:   23/30    Loss: 2.94229896068573

Epoch:   23/30    Loss: 3.069216561317444

Epoch:   23/30    Loss: 3.1417745113372804

Epoch:   23/30    Loss: 3.2345378160476685

Epoch:   23/30    Loss: 3.0616297245025637

Epoch:   23/30    Loss: 3.377500891685486

Epoch:   23/30    Loss: 3.0858586549758913

Epoch:   23/30    Loss: 3.179185390472412

Epoch:   23/30    Loss: 3.0009230852127073

Epoch:   23/30    Loss: 3.102643585205078

Epoch:   23/30    Loss: 3.089742231369018

Epoch:   23/30    Loss: 3.091712808609009

Epoch:   23/30    Loss: 3.3067859649658202

Epoch:   23/30    Loss: 3.1629625797271728

Epoch:   23/30    Loss: 3.3508126974105834

Epoch:   23/30    Loss: 3.0380665063858032

Epoch:   23/30    Loss: 3.266673445701599

Epoch:   23/30    Loss: 3.0461442470550537

Epoch:   23/30    Loss: 3.4735946893692016

Epoch:   23/30    Loss: 3.2324275970458984

Epoch:   23/30    Loss: 3.318776822090149

Epoch:   23/30    Loss: 3.1580111503601076

Epoch:   23/30    Loss: 3.2615506649017334

Epoch:   23/30    Loss: 3.204213285446167

Epoch:   23/30    Loss: 3.185917115211487

Epoch:   23/30    Loss: 3.282806468009949

Epoch:   23/30    Loss: 2.9957367181777954

Epoch:   23/30    Loss: 3.121064019203186

Epoch:   23/30    Loss: 3.177745318412781

Epoch:   23/30    Loss: 3.3922178983688354

Epoch:   23/30    Loss: 3.2619096517562864

Epoch:   23/30    Loss: 3.3655827045440674

Epoch:   23/30    Loss: 3.1131542682647706

Epoch:   23/30    Loss: 3.21880414485931

Epoch:   24/30    Loss: 3.135641312599182

Epoch:   24/30    Loss: 3.24179584980011

Epoch:   24/30    Loss: 3.176082491874695

Epoch:   24/30    Loss: 3.2981722831726072

Epoch:   24/30    Loss: 3.360953116416931

Epoch:   24/30    Loss: 3.1519974946975706

Epoch:   24/30    Loss: 3.015295147895813

Epoch:   24/30    Loss: 3.1989445447921754

Epoch:   24/30    Loss: 2.9170986652374267

Epoch:   24/30    Loss: 3.232534623146057

Epoch:   24/30    Loss: 3.3804799795150755

Epoch:   24/30    Loss: 3.2485100269317626

Epoch:   24/30    Loss: 3.1997549533843994

Epoch:   24/30    Loss: 3.328173565864563

Epoch:   24/30    Loss: 3.2600611448287964

Epoch:   24/30    Loss: 3.066431427001953

Epoch:   24/30    Loss: 3.360077166557312

Epoch:   24/30    Loss: 3.1629889249801635

Epoch:   24/30    Loss: 3.285318064689636

Epoch:   24/30    Loss: 3.112757110595703

Epoch:   24/30    Loss: 3.2252721548080445

Epoch:   24/30    Loss: 3.210455393791199

Epoch:   24/30    Loss: 3.301438403129578

Ep

Epoch:   25/30    Loss: 3.2209051132202147

Epoch:   25/30    Loss: 3.134887528419495

Epoch:   25/30    Loss: 3.1853341817855836

Epoch:   25/30    Loss: 3.246865248680115

Epoch:   25/30    Loss: 3.0828939199447634

Epoch:   25/30    Loss: 3.0018104553222655

Epoch:   25/30    Loss: 3.0936338663101197

Epoch:   25/30    Loss: 3.365938496589661

Epoch:   25/30    Loss: 3.1313601970672607

Epoch:   25/30    Loss: 3.0532853841781615

Epoch:   25/30    Loss: 2.898368167877197

Epoch:   25/30    Loss: 3.0684903144836424

Epoch:   25/30    Loss: 3.104542064666748

Epoch:   25/30    Loss: 3.1932831525802614

Epoch:   25/30    Loss: 2.861685800552368

Epoch:   25/30    Loss: 3.0869449853897093

Epoch:   25/30    Loss: 3.138780927658081

Epoch:   25/30    Loss: 3.0230174541473387

Epoch:   25/30    Loss: 3.2426592826843263

Epoch:   25/30    Loss: 3.1913164138793944

Epoch:   25/30    Loss: 3.1021681785583497

Epoch:   25/30    Loss: 3.0600795984268188

Epoch:   25/30    Loss: 3.1615015506744

Epoch:   26/30    Loss: 3.0487446069717405

Epoch:   26/30    Loss: 3.116301345825195

Epoch:   26/30    Loss: 3.1632114171981813

Epoch:   26/30    Loss: 3.1501529693603514

Epoch:   26/30    Loss: 2.9575628519058226

Epoch:   26/30    Loss: 2.9804103136062623

Epoch:   26/30    Loss: 3.1088504076004027

Epoch:   26/30    Loss: 3.0774781465530396

Epoch:   26/30    Loss: 2.988922095298767

Epoch:   26/30    Loss: 3.024422812461853

Epoch:   26/30    Loss: 3.1437524795532226

Epoch:   26/30    Loss: 3.1424004793167115

Epoch:   26/30    Loss: 3.2371719598770143

Epoch:   26/30    Loss: 3.039684510231018

Epoch:   26/30    Loss: 3.1193310260772704

Epoch:   26/30    Loss: 3.0099663972854613

Epoch:   26/30    Loss: 2.823675513267517

Epoch:   26/30    Loss: 2.953162693977356

Epoch:   26/30    Loss: 2.999940538406372

Epoch:   26/30    Loss: 3.2268234968185423

Epoch:   26/30    Loss: 3.112960624694824

Epoch:   26/30    Loss: 2.968586564064026

Epoch:   26/30    Loss: 3.060403919219971

### Question: How did you decide on your model hyperparameters? 
For example, did you try different sequence_lengths and find that one size made the model converge faster? What about your hidden_dim and n_layers; how did you decide on those?

**Answer:** I initially tried (sequence_length,batch size) combination of (100, 128) and (100, 64). I assummed a sequence length of 100 was long enough to extract context and short enough to ease computation. These values were with an embedding dimension of int(vocab_size ** 0.25), a formula which I later discovered was for only the skip-gram model. These combinations gave me a minimum loss of 5.42. Then, I decided to use an embedding dimension of 500 (similar to the table comparing embedding dimensions given in the classroom) still with a sequence length and batch size of 100 and 64 respectively and this gave me a minimum loss of 2.71 after 20 epochs. However, I noticed the loss was still decreasing and the format of the script I generated with this model was a bit off, so I increased the number of epochs to 30 which achieved a training loss of 1.78. At the start of setting the hyperparameters, I used 512 hidden dimension and 3 layers lstm (ensuring a lot of feature extraction)just to see how long it takes to train for 10 epochs. Since it didn't take a long period of time and I was already achieving a trainig loss less than 3.5(the maximum loss on the project rubic), I stuck with those values as I increased number of epochs.

---
# Checkpoint

After running the above training cell, your model will be saved by name, `trained_rnn`, and if you save your notebook progress, **you can pause here and come back to this code at another time**. You can resume your progress by running the next cell, which will load in our word:id dictionaries _and_ load in your saved model by name!

In [ ]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import torch
import helper
import problem_unittests as tests

_, vocab_to_int, int_to_vocab, token_dict = helper.load_preprocess()
trained_rnn = helper.load_model('./save/trained_rnn')

## Generate TV Script
With the network trained and saved, you'll use it to generate a new, "fake" Seinfeld TV script in this section.

### Generate Text
To generate the text, the network needs to start with a single word and repeat its predictions until it reaches a set length. You'll be using the `generate` function to do this. It takes a word id to start with, `prime_id`, and generates a set length of text, `predict_len`. Also note that it uses topk sampling to introduce some randomness in choosing the most likely next word, given an output set of word scores!

In [ ]:
"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
import torch.nn.functional as F

def generate(rnn, prime_id, int_to_vocab, token_dict, pad_value, predict_len=100):
    """
    Generate text using the neural network
    :param decoder: The PyTorch Module that holds the trained neural network
    :param prime_id: The word id to start the first prediction
    :param int_to_vocab: Dict of word id keys to word values
    :param token_dict: Dict of puncuation tokens keys to puncuation values
    :param pad_value: The value used to pad a sequence
    :param predict_len: The length of text to generate
    :return: The generated text
    """
    rnn.eval()
    
    # create a sequence (batch_size=1) with the prime_id
    current_seq = np.full((1, sequence_length), pad_value)
    current_seq[-1][-1] = prime_id
    predicted = [int_to_vocab[prime_id]]
    
    for _ in range(predict_len):
        if train_on_gpu:
            current_seq = torch.LongTensor(current_seq).cuda()
        else:
            current_seq = torch.LongTensor(current_seq)
        
        # initialize the hidden state
        hidden = rnn.init_hidden(current_seq.size(0))
        
        # get the output of the rnn
        output, _ = rnn(current_seq, hidden)
        
        # get the next word probabilities
        p = F.softmax(output, dim=1).data
        if(train_on_gpu):
            p = p.cpu() # move to cpu
         
        # use top_k sampling to get the index of the next word
        top_k = 5
        p, top_i = p.topk(top_k)
        top_i = top_i.numpy().squeeze()
        
        # select the likely next word index with some element of randomness
        p = p.numpy().squeeze()
        word_i = np.random.choice(top_i, p=p/p.sum())
        
        # retrieve that word from the dictionary
        word = int_to_vocab[word_i]
        predicted.append(word)     
        
        # the generated word becomes the next "current sequence" and the cycle can continue
        current_seq = np.roll(current_seq, -1, 1)
        current_seq[-1][-1] = word_i
    
    gen_sentences = ' '.join(predicted)
    
    # Replace punctuation tokens
    for key, token in token_dict.items():
        ending = ' ' if key in ['\n', '(', '"'] else ''
        gen_sentences = gen_sentences.replace(' ' + token.lower(), key)
    gen_sentences = gen_sentences.replace('\n ', '\n')
    gen_sentences = gen_sentences.replace('( ', '(')
    
    # return all the sentences
    return gen_sentences

### Generate a New Script
It's time to generate the text. Set `gen_length` to the length of TV script you want to generate and set `prime_word` to one of the following to start the prediction:
- "jerry"
- "elaine"
- "george"
- "kramer"

You can set the prime word to _any word_ in our dictionary, but it's best to start with a name for generating a TV script. (You can also start with any other names you find in the original text file!)

In [ ]:
# run the cell multiple times to get different results!
gen_length = 500 # modify the length to your preference
prime_word = "george" # name for starting the script

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
pad_word = helper.SPECIAL_WORDS['PADDING']
generated_script = generate(trained_rnn, vocab_to_int[prime_word + ':'], int_to_vocab, token_dict, vocab_to_int[pad_word], gen_length)
print(generated_script)

#### Save your favorite scripts

Once you have a script that you like (or find interesting), save it to a text file!

In [24]:
# save script to a text file
f =  open("generated_script_1.txt","w")
f.write(generated_script)
f.close()

# The TV Script is Not Perfect
It's ok if the TV script doesn't make perfect sense. It should look like alternating lines of dialogue, here is one such example of a few generated lines.

### Example generated script

>jerry: what about me?
>
>jerry: i don't have to wait.
>
>kramer:(to the sales table)
>
>elaine:(to jerry) hey, look at this, i'm a good doctor.
>
>newman:(to elaine) you think i have no idea of this...
>
>elaine: oh, you better take the phone, and he was a little nervous.
>
>kramer:(to the phone) hey, hey, jerry, i don't want to be a little bit.(to kramer and jerry) you can't.
>
>jerry: oh, yeah. i don't even know, i know.
>
>jerry:(to the phone) oh, i know.
>
>kramer:(laughing) you know...(to jerry) you don't know.

You can see that there are multiple characters that say (somewhat) complete sentences, but it doesn't have to be perfect! It takes quite a while to get good results, and often, you'll have to use a smaller vocabulary (and discard uncommon words), or get more data.  The Seinfeld dataset is about 3.4 MB, which is big enough for our purposes; for script generation you'll want more than 1 MB of text, generally. 

# Submitting This Project
When submitting this project, make sure to run all the cells before saving the notebook. Save the notebook file as "dlnd_tv_script_generation.ipynb" and save another copy as an HTML file by clicking "File" -> "Download as.."->"html". Include the "helper.py" and "problem_unittests.py" files in your submission. Once you download these files, compress them into one zip file for submission.